site/en/docs/extensions/reference/declarativeNetRequest/index.md

```diff
@@ -158,102 +150,139 @@ examples of URL filters:
+An extension can specify up to **50** static [rulesets](#type-Ruleset) as part of the `"rule_resources"` manifest key, but only **10** of these rulesets can be enabled at a time. The latter is called the [`MAX_NUMBER_OF_ENABLED_STATIC_RULESETS`](#property-MAX_NUMBER_OF_ENABLED_STATIC_RULESETS). Collectively, those rulesets are guaranteed at least **30,000** rules. This is called the [`GUARANTEED_MINIMUM_STATIC_RULES`](#property-GUARANTEED_MINIMUM_STATIC_RULES). The number of rules available after that depends on how many rules are enabled by all the extensions installed on a user's browser. You can find this number at runtime by calling [`getAvailableStaticRuleCount()`](#method-getAvailableStaticRuleCount). You can see [an example of this](#update-static-rulesets) under [code examples](#code-examples).
 
-When rules are applied to browsers with pages in the service worker's cached storage, the browser may ignore the set rule for those specific pages until the cached storage is cleared. This is because cached storage is intended to be persistent, and many features like offline use do not expect the cache to be cleared without also clearing a service worker's registration as well. For cases when extensions utilizing declarativeNetRequest must be enabled and disabled repeatedly, the [`chrome.browsingData`](/docs/extensions/reference/browsingData/) API may be used to clear the cache to guarantee proper functionality.
+#update-static-rulesets
 
-## Example {: #example }
+#### Dynamic and session rules {: #dynamic-session-rules }
 
-{% Label %}manifest.json:{% endLabel %}
+The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed **5000**. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`][#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES].
 
-```json
-{
-  "name": "declarativeNetRequest extension",
-  "version": "1",
-  "declarative_net_request": {
-    "rule_resources": [{
-      "id": "ruleset_1",
-      "enabled": true,
-      "path": "rules.json"
-    }]
-  },
-  "permissions": [
-    "declarativeNetRequest"
-  ],
-  "host_permissions": [
-    "*://*.google.com/*",
-    "*://*.abcd.com/*",
-    "*://*.example.com/*",
-    "https://*.xyz.com/*",
-    "*://*.headers.com/*",
-  ],
-  "manifest_version": 3
+#### Rules that use regex {: #regex-rules }
+
+All types of rules can use regular expressions; however the total number of regex rules, of each type cannot exceed **1000**. This is called the [MAX_NUMBER_OF_REGEX_RULES](#property-MAX_NUMBER_OF_REGEX_RULES).
+
+### Interactions with service workers {: #interact-w-service-workers }
+
+Requests and responses intercepted by the `onfetch` handler in a host page's service worker are not affected by declarativeNetRequet. Consequently such service workers can make it appear as though a rule doesn't work. To properly test your rules, you may need to disable or remove one or more of the service workers in your installation of Chrome.
```

Comment:
Do I understand this right that publishers could simply handle every request in the `onfetch` handler to avoid ad blockers functioning on their pages?

Reply:
So I was trying to avoid opening this can of worms and I think the above text is still vague enough to sidestep this - but if you call fetch in a service worker, those requests are affected by DNR. So even if a web service worker intercepts a request, the only way it can meaningfully respond is (a) if it already has a response cached or (b) if it does a fetch to make the request anyway, and DNR will apply. Hopefully that makes sense.

Reply:
We need to add this detail otherwise it is very misleading.

Reply:
That opens another can of worms. If I understand what you're saying correctly, whether and how dNR works depends on which [caching algorithm](https://web.dev/offline-cookbook/#serving-suggestions) a host page uses.

- *Cache only*- dNR will always fail
- *Network only*- dNR will always succeed
- *Cache, falling back to network*- dNR will fail when the item is in the cache, succeed otherwise
- *Cache and network race*- Who knows what dNR will do?
- *Network falling back to cache*- Whenever there's an item in the cache, dNR will appear to fail, but succeed otherwise.
- *Generic fallback*- dNR will appear to fail when the network can't be reached.
- *Service worker templating*- dNR will appear to fail.

I can't see how an extension developer would design around this.

Reply:
There must be some very good reason why dNR doesn't affect items retrieved from `CacheStorage`.

Reply:
@jpmedley - all of the above sounds right 👍 

I don't feel quite as pessimistic about being able to document this. How about the following (changes in bold)?

> Requests and responses intercepted by the `onfetch` handler in a host page's service worker are not affected by declarativeNetRequest. Consequently such service workers can make it appear as though a rule doesn't work. To properly test your rules, you may need to disable or remove one or more of the service workers in your installation of Chrome. **Note that DNR does apply if the service worker goes on to make subsequent requests to the network, even if these are to the same path that was previously intercepted.**

On your other point:

> I can't see how an extension developer would design around this.

I think this depends on the developer's goal. If you're making something to block trackers, a request to /track_user that never reaches the network is useless, so the developers of those websites aren't going to enable any caching. In that case you can probably just ignore all of this.

If you are trying to make something to block a page from loading, you have bigger problems than this. DNR doesn't have any affect on single page apps that use `history.pushState` to change what is showing in the address bar for example.

So TLDR, I think making developers aware of this gotcha is important, but I don't expect it actually causes many issues in practice.

Reply:
I don't thinks this phrasing is correct on it's own:

> Requests and responses intercepted by the `onfetch` handler in a host page's service worker are not affected by declarativeNetRequest. 


Because they are affected, if the sw makes a subsequent fetch request. I still like Oliver's original phrasing, we could add an example to make it more clear:

```
DNR only applies to requests that reach the network stack. This does include responses from the HTTP cache, but does not include responses generated by a website's service worker. For example, if a service worker intercepts a request in an `onFetch` handler and returns a cached or generated response, the response will not be affected by your DNR rules. However, if the service worker makes a subsequent `fetch` request, your DNR rules will apply to this new request.
```